README.md

```diff
@@ -0,0 +1,226 @@
+
+Rails creates a signature for every cookie it sets, and appends the signature to the cookie.
+
+When it receives a cookie, Rails verifies that the signature matches the content (that is, that `sign(cookie_body, secret_key_base) == cookie_signature`).
+
+This prevents cookie tampering. If a user tries to edit their cookie and change the `cart_id`, they signature won't match, and Rails will silently ignore the cookie and set a new one.
+
+Cryptography is a deep rabbit hole. At this point, you don't need to worry about the specifics of how cryptography works, just that Rails and other frameworks use it to ensure that session data which is set on the server can't be edited by users.
+
+# Tying it together
+
+In our `items_controller.rb`, we might have an `add_to_cart` method, which is called
+when the user adds something to their cart. It might work something like this:
+
+```ruby
+# Routed from POST /items/:id/add_to_cart
+def add_to_cart
+  # Get the item from the path
+  @item = Item.find(params[:id])
+  
+  # Load the cart from the session, or create a new empty cart.
+  cart = session[:cart] || []
+  cart << @item.id
+
+  # Save the cart in the session.
+  session[:cart] = cart
+end
+```
+
+That's it! It's common to wrap this up in a helper method:
+
+```ruby
+class ApplicationController < ActionController::Base
+  def current_cart
+    session[:cart] ||= []
+  end
+end
+```
+
+So now our controller looks like this:
+
+```ruby
+# Routed from POST /items/:id/add_to_cart
+def add_to_cart
+  # Get the item from the path
+  @item = Item.find(params[:id])
+  
+  # Load the cart from the session, or create a new empty cart.
+  current_cart << @item.id
+end
+```
+
+This way, we can use `current_cart` in our views and layouts too. For example, we may want to show the user how many items are in their cart as part of the layout.
+
+## Conclusion
+
+Cookies are foundational for the modern web.
+
+Most sites use cookies, either to let their users log in, or to keep track of their
+shopping carts, or record other ephemeral session data. Almost nobody thinks these are bad uses of cookies: nobody really believes that you should have to type in your username and password on every page, or that your shopping cart should clear if you reload the page.
+
+But cookies just let you store data in a user's browser, so by nature, they can be used for more controversial endeavors. 
+
+For example, Google AdWords sets a cookie and uses that cookie to track what ads you've seen and which ones you've clicked on. The tracking information helps AdWords decide what ads to show you.
+
+This is why, if you click on an ad, you may find that the ad follows you around the Internet. It turns out that this behavior is as effective as it is annoying: people are far more likely to buy things from ads that they've clicked on once.
+
+This use of cookies worries people. And so now the EU has a law.
```

Comment:
maybe link to this: https://en.wikipedia.org/wiki/HTTP_cookie#EU_cookie_directive